### PR TITLE
Dialogue tasks unit test

### PR DESCRIPTION
**Pull request by Zhilin123**
# What does this PR do ?

Add unit test to dialogue tasks and some refactor to facilitate unit test

**Collection**: NLP

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/data/dialogue/data_processor/data_processor.py [hunk 1]
```diff
@@ -50,3 +53,29 @@ def get_dev_examples(self):
...
+            n_dev = int(n_samples * (dev_proportion / 100))
+            dev_idxs = random.sample(list(range(n_samples)), n_dev)
+            if dataset_split == "dev":
+                idxs = dev_idxs
+            else:
+                dev_idxs_set = set(dev_idxs)
+                train_idxs = [idx for idx in list(range(n_samples)) if idx not in dev_idxs_set]
+                idxs = train_idxs
+
+        elif dataset_split == "test":
```
Comment (okuchaiev) [inline nemo/collections/nlp/data/dialogue/data_processor/data_processor.py:78]: please make sure every
```
if ...
elif ...
else
``` 
block ends with ``else``. In this specific case, in else you'd raise an exception and tell the user supported names of splits (train, dev, test)

File: nemo/collections/nlp/data/dialogue/data_processor/data_processor.py [hunk 2]
```diff
@@ -50,3 +53,29 @@ def get_dev_examples(self):
...
+
+    @staticmethod
+    def get_relevant_idxs(dataset_split, n_samples, dev_proportion):
+        """
+        Obtain indexes for each dataset_split, when train and dev sets are not in separate files
+        
+        Args: 
+            dataset_split: train, dev or test
+            n_samples: total number of samples
+            dev_proportion: value from 1 to 99 that represent proportion of data in dev set
```
Comment (okuchaiev) [inline nemo/collections/nlp/data/dialogue/data_processor/data_processor.py:65]: Can you please document "Returns:" of this method too?

File: nemo/collections/nlp/data/dialogue/dataset/dialogue_gpt_classification_dataset.py [hunk 3]
```diff
@@ -120,7 +110,9 @@ def default_encode(self, sentence):
         attn_masks = torch.squeeze(encodings_dict['attention_mask'])
         return encodings_dict, input_ids, attn_masks
 
-    def linearize_slots(self, slots):
+    @staticmethod
+    def linearize_slots(slots):
+
```
Comment (okuchaiev) [inline nemo/collections/nlp/data/dialogue/dataset/dialogue_gpt_classification_dataset.py:115]: can you please add docstring here?

File: nemo/collections/nlp/data/dialogue/dataset/dialogue_gpt_classification_dataset.py [hunk 4]
```diff
@@ -71,19 +70,10 @@ def __init__(self, dataset_split: str, dialogues_processor: object, tokenizer, c
         self.lm_features = self.get_lm_samples()
 
     def transform(self, label):
```
Comment (okuchaiev) [inline nemo/collections/nlp/data/dialogue/dataset/dialogue_gpt_classification_dataset.py:72]: can you please add docstring here

File: nemo/collections/nlp/data/dialogue/dataset/dialogue_s2s_generation_dataset.py [hunk 5]
```diff
@@ -42,21 +42,22 @@ def __init__(self, dataset_split: str, dialogues_processor: object, tokenizer, c
         if self.cfg.debug_mode:
             self.features = self.features[:16]
 
-    def format_actions(self, actions):
+    @staticmethod
+    def format_actions(prompt_template, actions):
```
Comment (okuchaiev) [inline nemo/collections/nlp/data/dialogue/dataset/dialogue_s2s_generation_dataset.py:46]: please add docstring here.

File: nemo/collections/nlp/data/dialogue/dataset/dialogue_s2s_generation_dataset.py [hunk 6]
```diff
@@ -42,21 +42,22 @@ def __init__(self, dataset_split: str, dialogues_processor: object, tokenizer, c
...
         actions_str = []
         for action in actions:
             act = action['act'].lower()
             slot = action['slot']
             value = action['values'][0] if action['values'] else ''
 
-            if self.cfg.prompt_template == 'values':
+            if prompt_template == 'values':
```
Comment (okuchaiev) [inline nemo/collections/nlp/data/dialogue/dataset/dialogue_s2s_generation_dataset.py:53]: same comment - please make sure this if eventually ends with "else"

**Review by okuchaiev — CHANGES_REQUESTED** [review]
Two small comments:
1) Please make sure every if eventually ends with else
2) Add docstrings for new public methods

File: nemo/collections/nlp/data/dialogue/data_processor/data_processor.py (hunk 1 above)
Comment (Zhilin123) [reply nemo/collections/nlp/data/dialogue/data_processor/data_processor.py:78]: sure

File: nemo/collections/nlp/data/dialogue/data_processor/data_processor.py (hunk 2 above)
Comment (Zhilin123) [reply nemo/collections/nlp/data/dialogue/data_processor/data_processor.py:65]: done!

File: nemo/collections/nlp/data/dialogue/dataset/dialogue_gpt_classification_dataset.py (hunk 4 above)
Comment (Zhilin123) [reply nemo/collections/nlp/data/dialogue/dataset/dialogue_gpt_classification_dataset.py:72]: sure

File: nemo/collections/nlp/data/dialogue/dataset/dialogue_gpt_classification_dataset.py (hunk 3 above)
Comment (Zhilin123) [reply nemo/collections/nlp/data/dialogue/dataset/dialogue_gpt_classification_dataset.py:115]: sure

File: nemo/collections/nlp/data/dialogue/dataset/dialogue_s2s_generation_dataset.py (hunk 6 above)
Comment (Zhilin123) [reply nemo/collections/nlp/data/dialogue/dataset/dialogue_s2s_generation_dataset.py:53]: sure!

File: nemo/collections/nlp/data/dialogue/dataset/dialogue_s2s_generation_dataset.py (hunk 5 above)
Comment (Zhilin123) [reply nemo/collections/nlp/data/dialogue/dataset/dialogue_s2s_generation_dataset.py:46]: sure

**Review by okuchaiev — APPROVED** [review]
lgtm

**Pull request merged**